File: sdks/typescript/src/types/realtime.ts
export interface RealtimeSnapshot<T = unknown> {
  type: "snapshot";
  id: string;
  topic: string;
  payload: T;
  generated_at: string;
}

export interface RealtimeEvent<T = unknown> {
  type: "event";
  id: string;
  topic: string;
  payload: T;
  generated_at: string;
}

export interface RealtimeError {
  type: "error";
  id?: string;
  topic?: string;
  message: string;
  generated_at: string;
}

export interface RealtimeComplete {
  type: "complete";
  id: string;
  topic: string;
  generated_at: string;
}

export interface RealtimeControlMessage {
  type: "subscribed" | "heartbeat" | "pong";
  id?: string;
  topic?: string;
  generated_at?: string;
}

export type RealtimeMessage<T = unknown> =
  | RealtimeSnapshot<T>
  | RealtimeEvent<T>
  | RealtimeError
  | RealtimeComplete
  | RealtimeControlMessage;

export interface ProgressEvent {
  step: string;
  message: string;
  progress: number;
  document_id?: string;
  status?: string;
  detail?: Record<string, unknown>;
}

File: sdks/python/src/bigrag/types/realtime.py
from __future__ import annotations

from typing import Any, Literal, NotRequired, TypedDict


class RealtimeSnapshot(TypedDict):
    type: Literal["snapshot"]
    id: str
    topic: str
    payload: Any
    generated_at: str


class RealtimeEvent(TypedDict):
    type: Literal["event"]
    id: str
    topic: str
    payload: Any
    generated_at: str


class RealtimeError(TypedDict):
    type: Literal["error"]
    message: str
    id: NotRequired[str]
    topic: NotRequired[str]
    generated_at: str


class RealtimeComplete(TypedDict):
    type: Literal["complete"]
    id: str
    topic: str
    generated_at: str


class RealtimeControlMessage(TypedDict):
    type: Literal["subscribed", "heartbeat", "pong"]
    id: NotRequired[str]
    topic: NotRequired[str]
    generated_at: NotRequired[str]


RealtimeMessage = (
    RealtimeSnapshot | RealtimeEvent | RealtimeError | RealtimeComplete | RealtimeControlMessage
)


class ProgressEvent(TypedDict):
    step: str
    message: str
    progress: float
    document_id: NotRequired[str]
    status: NotRequired[str]
    detail: NotRequired[dict[str, Any]]
